staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -124,38 +130,85 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
 	}
 
 	// Make sure the necessary fields exist
-	pv.Spec.NodeAffinity = new(v1.VolumeNodeAffinity)
-	pv.Spec.NodeAffinity.Required = new(v1.NodeSelector)
-	pv.Spec.NodeAffinity.Required.NodeSelectorTerms = make([]v1.NodeSelectorTerm, 1)
+	if pv.Spec.NodeAffinity == nil {
+		pv.Spec.NodeAffinity = new(v1.VolumeNodeAffinity)
+	}
+
+	if pv.Spec.NodeAffinity.Required == nil {
+		pv.Spec.NodeAffinity.Required = new(v1.NodeSelector)
+	}
+
+	if len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) == 0 {
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms = make([]v1.NodeSelectorTerm, 1)
+	}
 
 	topology := v1.NodeSelectorRequirement{
 		Key:      topologyKey,
 		Operator: v1.NodeSelectorOpIn,
 		Values:   zones,
 	}
 
-	pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions = append(
-		pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions,
-		topology,
-	)
+	// add the CSI topology to each term
+	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions = append(
+			pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions,
+			topology,
+		)
+	}
 
 	return nil
 }
 
-// translateTopology converts existing zone labels or in-tree topology to CSI topology.
+// removeTopology removes the topology from the given PV. Return false
+// if the topology key is not found
+func removeTopology(pv *v1.PersistentVolume, topologyKey string) bool {
+	// Make sure the necessary fields exist
+	if pv == nil || pv.Spec.NodeAffinity == nil || pv.Spec.NodeAffinity.Required == nil ||
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms == nil || len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) == 0 {
+		return false
+	}
+
+	succeed := false
+	for termIndex, nodeSelectorTerms := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		nsrequirements := nodeSelectorTerms.MatchExpressions
+
+		index := -1
+		for i, nodeSelectorRequirement := range nsrequirements {
+			if nodeSelectorRequirement.Key == topologyKey {
+				index = i
+				break
```

Comment:
okay, I changed the logic to remove the key from:
1. Multiple terms(This will be the regional PD case I think)
2. Single term with duplicate key match expression(although I doubt this will happen)

Also add removeTopology test